File: laptop-store-backend/controller/src/main/java/org/example/handler/GlobalExceptionHandler.java
package org.example.handler;

import com.fasterxml.jackson.core.JsonProcessingException;
import org.example.exception.InvalidCredentialException;
import org.example.util.TranslatorUtil;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;

public class GlobalExceptionHandler {
    @ExceptionHandler(value = InvalidCredentialException.class)
    public ResponseEntity<?> handleAuthenticationException(InvalidCredentialException e) {
        String message = TranslatorUtil.toLocale("invalid_credential_message");
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(message);
    }

    @ExceptionHandler(value = IllegalArgumentException.class)
    public ResponseEntity<?> handleIllegalArgumentException(IllegalArgumentException e) {
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(e.getMessage());
    }

    @ExceptionHandler(value = MethodArgumentNotValidException.class)
    public ResponseEntity<?> handleServerException(MethodArgumentNotValidException e) {
        String message = e.getBindingResult().getAllErrors().get(0).getDefaultMessage();
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(message);
    }

    @ExceptionHandler(value = {JsonProcessingException.class, Exception.class})
    public ResponseEntity<?> handleServerException(Exception e) {
        e.printStackTrace();
        String message = TranslatorUtil.toLocale("server_error_message");
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(message);
    }
}

File: laptop-store-backend/controller/src/main/kotlin/org/example/handler/GlobalExceptionHandler.kt
package org.example.handler

import com.fasterxml.jackson.core.JsonProcessingException
import org.example.constant.ErrorMessageConstants
import org.example.exception.InvalidCredentialException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import java.lang.Exception

@ControllerAdvice
class GlobalExceptionHandler {
    @ExceptionHandler(value = [InvalidCredentialException::class])
    fun handleAuthenticationException(e: InvalidCredentialException): ResponseEntity<String> {
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(ErrorMessageConstants.INVALID_CREDENTIAL)
    }

    @ExceptionHandler(value = [IllegalArgumentException::class])
    fun handleIllegalArgumentException(e: IllegalArgumentException): ResponseEntity<String> {
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(e.message)
    }

    @ExceptionHandler(value = [MethodArgumentNotValidException::class])
    fun handleServerException(e: MethodArgumentNotValidException): ResponseEntity<String> {
        val message = e.bindingResult.allErrors[0].defaultMessage
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(message)
    }

    @ExceptionHandler(value = [JsonProcessingException::class, Exception::class])
    fun handleServerException(e: Exception): ResponseEntity<String> {
        e.printStackTrace()
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(ErrorMessageConstants.SERVER_ERROR)
    }
}